Source/DafnyTestGeneration/README.md

```diff
@@ -0,0 +1,149 @@
+- [ProgramModifier.cs](ProgramModifier.cs) - an abstract class that contains utilities for annotating Boogie Programs, inlining Dafny methods after they are translated to Boogie, etc. This is all necessary for generating instances of `ProgramModification`
+- [BlockBasedModifier.cs](BlockBasedModifier.cs) and [PathBasedModifier.cs](PathBasedModifier.cs) - subclasses of `ProgramModifier` that generate instances of `ProgramModification` by inserting assertions that fail when a particular block is visited or a particular path is taken, respectively.
+- [DafnyInfo.cs](DafnyInfo.cs) - takes a Dafny Program and extracts certain information about it that is then used in `TestMethod.cs` and `Main.cs`
+
+## How to Generate Tests?
+
+- Test generation currently works with all basic types, user-defined classes, sequences, sets, and maps. It does not work with datatypes, arrays, and multisets. It is also not possible to generate tests for constructors.
+- To generate block- or path-coverage tests use the `/testMode:Block` or `/testMode:Path` arguments respectively. You will likely also need `definiteAssignment:3`, which will ensure that you generate tests even for those methods that do not contain any assertions or pre-\post-conditions.
+- If you wish to test a particular method rather than all the methods in a file, you can specify such a method with the `/testTargetMethod` command line argument and providing the fully qualified method name.
+- If you are using `/testTargetMethod` and would like to inline methods that are called from the method of interest, you can do so by setting `/testInlineDepth` to something larger than zero (zero is the default). The `/verifyAllModules` argument might also be relevant if the methods to be inlined are defined in included files.
+- To deal with loops, you should use `/loopUnroll` and also `/testSeqLengthLimit`. The latter argument adds an axiom that limits the length of any sequence to be no greater than some specified value. This restriction can be used to ensure that the number of loop unrolls is sufficient with respect to the length of any input sequence but it can also cause the program to miss certain corner cases.
+- The`/testMode:DeadCode` argument will make Dafny identify potential dead code in the specified file. Note that false negatives are possible if `/loopUnroll` is not used. False positives are also possible for a variety of reasons, such as `/loopUnroll` being assigned not high enough value.
+
+## How to Run Tests?
+
+To run the tests, you would first need to compile the code to some other language. Once this is done, the main difficulty is to mock the arguments that the test methods take. The following code (to be inserted into the class with the tests) should do the job in Java:
```

Comment:
Is it possible to add another example that ends up producing test methods that take arguments? It's a bit hard to follow this since the example immediately below doesn't.

Reply:
Changed the example in one of my latest commits to show how to work around test methods that take arguments.